### ISSUE TITLE
cmd/compile: slicing rules too restrictive for generic slice operands

### ISSUE THREAD

**Issue by griesemer** [golang/go#49566]
Reminder issue.

Currently not permitted (no structural type):
```Go
func _[S []byte|string](s S, i, j int) S {
   return s[i : j]
}
```

A little bit tricky to get right because we sometimes have to preserve the string type.

**Comment by go101** [comment]
Is this the same problem?
```Go
func Entry[T []int | map[int]int](c T, i int) int {
	return c[i] // invalid operation: cannot index c (variable of type T constrained by sliceOrMap
}
```

**Comment by griesemer** [comment]
Not quite the same problem (simple indexing rules are more relaxed) but similar: we don't allow mixing maps with non-map types for indexing. Maps just behave very differently than slices. Also, we don't have strong reasons (yet) that suggest that this must be permitted.

We are aware that per the generics proposal this could or even "should" work.  But the generics proposal is not very specific with the rule details. For 1.18 and the spec (in the process of being written) we need to be very specific.

In short, for 1.18 we play it somewhat conservative. It's easy to relax the rules down the road, but impossible to change if we make a mistake by being overly broad. Thanks.

**Comment by go101** [comment]
The proposal does provide such an example, but I think now the proposal has been outdated.

OK, I will submit a proposal to support this.
